checkpoint/handler.go

```diff
@@ -32,110 +32,111 @@ func NewHandler(k Keeper, contractCaller helper.IContractCaller) sdk.Handler {
+	logger.Debug("DividendAccounts of all validators", "dividendAccountsLength", len(dividendAccounts))
+
+	// Get account root has from dividend accounts
 	accountRoot, err := types.GetAccountRootHash(dividendAccounts)
-	k.Logger(ctx).Info("Validator Account root hash generated", "AccountRootHash", hmTypes.BytesToHeimdallHash(accountRoot).String())
+	if err != nil {
+		logger.Error("Error while fetching account root hash", "error", err)
+		return common.ErrBadBlockDetails(k.Codespace()).Result()
+	}
+	logger.Debug("Validator account root hash generated", "accountRootHash", hmTypes.BytesToHeimdallHash(accountRoot).String())
 
+	// Compare stored root hash to msg root hash
 	if !bytes.Equal(accountRoot, msg.AccountRootHash.Bytes()) {
-		k.Logger(ctx).Error("AccountRootHash of current state", hmTypes.BytesToHeimdallHash(accountRoot).String(),
-			"doesn't match with AccountRootHash of msg", msg.AccountRootHash)
+		logger.Error(
+			"AccountRootHash of current state doesn't match from msg",
+			"hash", hmTypes.BytesToHeimdallHash(accountRoot).String(),
+			"msgHash", msg.AccountRootHash,
+		)
 		return common.ErrBadBlockDetails(k.Codespace()).Result()
 	}
 
-	k.Logger(ctx).Debug("AccountRootHash matches")
+	//
+	// Validate proposer
+	//
 
-	// check proposer in message
-	if !bytes.Equal(msg.Proposer.Bytes(), k.sk.GetValidatorSet(ctx).Proposer.Signer.Bytes()) {
-		k.Logger(ctx).Error("Invalid proposer in message",
-			"currentProposer", k.sk.GetValidatorSet(ctx).Proposer.Signer.String(),
-			"checkpointProposer", msg.Proposer.String())
-		return common.ErrBadProposerDetails(k.Codespace(), k.sk.GetValidatorSet(ctx).Proposer.Signer).Result()
+	// Check proposer in message
+	validatorSet := k.sk.GetValidatorSet(ctx)
+	if validatorSet.Proposer == nil {
+		logger.Error("No proposer in validator set", "msgProposer", msg.Proposer.String())
+		return common.ErrInvalidMsg(k.Codespace(), "No proposer in stored validator set").Result()
 	}
-	k.Logger(ctx).Debug("Valid proposer in checkpoint")
-
-	// add checkpoint to buffer
-	// Add AccountRootHash to CheckpointBuffer
-	k.SetCheckpointBuffer(ctx, hmTypes.CheckpointBlockHeader{
-		StartBlock:      msg.StartBlock,
-		EndBlock:        msg.EndBlock,
-		RootHash:        msg.RootHash,
-		AccountRootHash: msg.AccountRootHash,
-		Proposer:        msg.Proposer,
-		TimeStamp:       timeStamp,
-	})
 
-	checkpoint, _ := k.GetCheckpointFromBuffer(ctx)
-	k.Logger(ctx).Debug("Adding good checkpoint to buffer to await ACK", "checkpointStored", checkpoint.String())
+	if !bytes.Equal(msg.Proposer.Bytes(), validatorSet.Proposer.Signer.Bytes()) {
+		logger.Error(
+			"Invalid proposer in msg",
+			"proposer", validatorSet.Proposer.Signer.String(),
+			"msgProposer", msg.Proposer.String(),
+		)
+		return common.ErrInvalidMsg(k.Codespace(), "Invalid proposer in msg").Result()
+	}
 
+	// Emit event for checkpoint
 	ctx.EventManager().EmitEvents(sdk.Events{
 		sdk.NewEvent(
 			types.EventTypeCheckpoint,
```

Comment:
Post checkpoint handler would be right place to emit this event as it confirms the state update.